### ISSUE TITLE
is that thread-safe ? e.g. while wallet generating ...

### ISSUE THREAD

**Issue by kostyll** [libbtc/libbtc#111]

**Comment by jonasschnelli** [comment]
It should be... but happy to see reports (and/or additional fixe pull requests)

**Comment by xloem** [comment]
I do not believe this codebase is thread safe at this time.  I do not see any locks in the source.  It does use libevent, which has thread safety properties, but these are not [enabled](http://www.wangafu.net/~nickm/libevent-book/Ref1_libsetup.html#_locks_and_threading) yet, and an interface respecting them does not appear to be exposed (I think that would imply the whole api would be driven by the event loop?  unsure).

I started a threaded app that used libbtc, and found it simplest to pump the event loop myself using libevent's `event_base_loop` call, passing libbtc's `nodegroup->event_base` structure member and the constant `EVLOOP_ONCE`.  When 0 is returned all clients have disconnected and the loop can terminate.  I lock around the call, and hold the same lock in asynchronous functions.  I expect this to make those functions thread-safe, but have not tested it too much at this time.

I imagine thread-safety would be implemented by adding a lock object to the structures, but I am not too experienced with the codebase.
